fuzz/fuzz_targets/ruff_formatter_validity.rs

```diff
@@ -0,0 +1,103 @@
+
+fn do_fuzz(case: &[u8]) -> Corpus {
+    // throw away inputs which aren't utf-8
+    let Ok(code) = std::str::from_utf8(case) else {
+        return Corpus::Reject;
+    };
+
+    // the settings are immutable to test_snippet, so we avoid re-initialising here
+    let linter_settings = SETTINGS.get_or_init(LinterSettings::default);
+    let format_options = PyFormatOptions::default();
+    ruff_linter::test::set_max_iterations(usize::MAX);
+
+    let linter_results = ruff_linter::linter::lint_only(
+        "fuzzed-source.py".as_ref(),
+        None,
+        &linter_settings,
+        Noqa::Enabled,
+        &SourceKind::Python(code.to_string()),
+        PySourceType::Python,
+        ParseSource::None,
+    );
+
+    if linter_results.error.is_some() {
+        return Corpus::Keep; // keep, but don't continue
+    }
+
+    let mut warnings = HashMap::new();
+
+    for msg in linter_results.data.0 {
+        let count: &mut usize = warnings.entry(msg.kind.name).or_default();
+        *count += 1;
+    }
+
+    // format the code once
+    if let Ok(formatted) = format_module_source(code, format_options.clone()) {
+        let formatted = formatted.as_code().to_string();
+
+        let linter_results = ruff_linter::linter::lint_only(
+            "fuzzed-source.py".as_ref(),
+            None,
+            &linter_settings,
+            Noqa::Enabled,
+            &SourceKind::Python(formatted.clone()),
+            PySourceType::Python,
+            ParseSource::None,
+        );
+
+        assert!(
+            linter_results.error.is_none(),
+            "formatter introduced a parse error"
+        );
+
+        for msg in linter_results.data.0 {
+            if let Some(count) = warnings.get_mut(&msg.kind.name) {
+                if let Some(decremented) = count.checked_sub(1) {
+                    *count = decremented;
+                } else {
+                    panic!(
+                        "formatter introduced additional linter warning: {msg:?}\ndiff: {}",
+                        TextDiff::from_lines(code, &formatted)
+                            .unified_diff()
+                            .header("Unformatted", "Formatted")
+                            .to_string()
+                    );
+                }
+            } else {
+                panic!(
+                    "formatter introduced new linter warning that was not previously present: {msg:?}\ndiff: {}",
+                    TextDiff::from_lines(code, &formatted)
+                        .unified_diff()
+                        .header("Unformatted", "Formatted")
+                        .to_string()
+                );
+            }
+        }
+    }
+
+    // unlike in the test framework, where the number of iterations is well-defined, we are only
+    // looking for situations where a fix is bad; thus, we set the iterations to "infinite"
+    let _ = ruff_linter::test::test_snippet(code, linter_settings);
```

Comment:
Can you explain the motivation for calling `test_snippet` here?

Reply:
See above.